### PR TITLE
fix: render dotted data keys whose root is missing

### LINKED ISSUE #288: Copy pasting elements with data key values to a new page breaks the page
If the new page does not have the client script for the data key (it is not needed in this case), then published page gives an error. 

Can be handled in two ways? Either strip the data keys while pasting to a new page, or, ignore the missing key, or, prompt the user to retain or strip the key if Client Script is missing. 

image

image

### PR DESCRIPTION
**Pull request by surajshetty3416**
Fixes #288

Blocks copied from one page to another keep their bindings, for example `hero.title`. When the target page's data script does not define `hero`, the raw key was emitted into the template and Jinja raised on the attribute read, so the whole page failed with a server error.

Dynamic values, dynamic props and visibility conditions each had their own copy of the key resolution. They now share `get_binding_key`, which guards a dotted path with the same `(a or {}).get('b', {})` chain that repeaters and component props already use. Flat keys stay raw so `0` and `""` still render as-is, and a key that is not a safe identifier renders nothing instead of broken Jinja.

For dynamic props the chain ends in `none` instead of `{}` and the template tests `is not none`, so a missing binding gets the prop's configured default while an empty object, `""`, `0`, `[]` and `False` pass through unchanged.

### CODE REVIEW

File: builder/builder/doctype/builder_page/builder_page.py [hunk 1]
```diff
@@ -952,21 +952,25 @@ def interpret_prop_value(prop_config: dict, data_key: dict | None) -> Any:
...
-	else:  # dataScript
-		if data_key:
-			base_key = extract_data_key(data_key)
-			key = jinja_safe_key(f"{base_key}.{prop_value}")
-		else:
-			key = prop_value
-
+	key = get_binding_key(prop_value, comes_from, data_key)
 	fallback = escape_single_quotes(default_value) if default_value is not None else "undefined"
 	return f"{{{{ {key} if {key} is defined else '{fallback}' }}}}"
```
[the remark this answers is not in the document]
Comment (surajshetty3416) [reply builder/builder/doctype/builder_page/builder_page.py]: Good catch. The props template now also requires the value to differ from `{}`, so a missing binding falls back to the configured default while `""`, `0`, `[]` and `False` still pass through. Fixed in 4ecdb468 with tests for both the missing and the resolved case.

File: builder/builder/doctype/builder_page/builder_page.py [hunk 2]
```diff
@@ -952,23 +952,28 @@ def interpret_prop_value(prop_config: dict, data_key: dict | None) -> Any:
...
-			base_key = extract_data_key(data_key)
-			key = jinja_safe_key(f"{base_key}.{prop_value}")
-		else:
-			key = prop_value
-
+	key = get_binding_key(prop_value, comes_from, data_key)
 	fallback = escape_single_quotes(default_value) if default_value is not None else "undefined"
-	return f"{{{{ {key} if {key} is defined else '{fallback}' }}}}"
+	# a guarded chain yields {} for a missing binding, which still counts as defined
+	return f"{{{{ {key} if {key} is defined and {key} != {{}} else '{fallback}' }}}}"
```
[the remark this answers is not in the document]
Comment (surajshetty3416) [reply builder/builder/doctype/builder_page/builder_page.py]: Agreed, comparing with `{}` conflated the two. The props chain now ends in `none` for a missing path (`jinja_safe_key(key, missing="none")`) and the template tests `is not none`, so an empty object, `""`, `0`, `[]` and `False` all pass through while a missing binding still gets its default. Fixed in a9150e78 with a test for the empty-object case; the repeater and dynamic-value chains keep `{}`.

File: builder/builder/doctype/builder_page/builder_page.py [hunk 3]
```diff
@@ -952,23 +952,28 @@ def interpret_prop_value(prop_config: dict, data_key: dict | None) -> Any:
...
-			base_key = extract_data_key(data_key)
-			key = jinja_safe_key(f"{base_key}.{prop_value}")
-		else:
-			key = prop_value
-
+	# props tell a missing path apart from an empty object, so the chain ends in none
+	key = get_binding_key(prop_value, comes_from, data_key, missing="none")
 	fallback = escape_single_quotes(default_value) if default_value is not None else "undefined"
-	return f"{{{{ {key} if {key} is defined else '{fallback}' }}}}"
+	return f"{{{{ {key} if {key} is defined and {key} is not none else '{fallback}' }}}}"
```
[the remark this answers is not in the document]
Comment (surajshetty3416) [reply builder/builder/doctype/builder_page/builder_page.py:976]: Leaving this as is. None already means "no value" for every other binding in Builder: dynamic values fall back on it (the `K or K in ['', 0]` test), visibility conditions hide on it, and Frappe returns None for any empty field. A prop bound to an empty field should show its configured default rather than pass null into the component, and a separate sentinel would make props the one binding that behaves differently. The prior `is defined` form only passed None through as a side effect of never checking the value at all.
